File: android/src/main/kotlin/africa/ejara/trustdart/coins/XTZ.kt
import africa.ejara.trustdart.Coin
import africa.ejara.trustdart.Numeric
import africa.ejara.trustdart.utils.base64String
import africa.ejara.trustdart.utils.toHex
import africa.ejara.trustdart.utils.toHexBytes
import africa.ejara.trustdart.utils.toLong
import com.google.protobuf.ByteString
import org.json.JSONObject
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.CoinType.TEZOS
import wallet.core.jni.HDWallet
import wallet.core.jni.Hash
import wallet.core.jni.proto.Tezos
import wallet.core.jni.proto.Tezos.*


class XTZ : Coin("XTZ", CoinType.TEZOS) {

    override fun getPublicKey(path: String, mnemonic: String, passphrase: String): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path).publicKeyEd25519.data().base64String()
    }

    override fun getRawPublicKey(path: String, mnemonic: String, passphrase: String): ByteArray? {
        val wallet = HDWallet(mnemonic, passphrase)
        return wallet.getKey(coinType, path)
            .publicKeyEd25519.data()
    }

    override fun signDataWithPrivateKey(
        path: String,
        mnemonic: String,
        passphrase: String,
        txData: String
    ): String? {
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val hash: ByteArray? = Hash.blake2b(txData.toHexBytes(), 32)
        return privateKey.sign(hash, coinType!!.curve()).toHex()
    }

    override fun signTransaction(
        path: String,
        txData: Map<String, Any>,
        mnemonic: String,
        passphrase: String
    ): String? {
        val cmd: String? = txData["cmd"] as String?
        val wallet = HDWallet(mnemonic, passphrase)
        val privateKey = wallet.getKey(coinType, path)
        val txHash: String?
        val publicKey = privateKey.publicKeyEd25519.data()
        val isRevealed = txData["isRevealed"] as Boolean?
        var revealOperation: Tezos.Operation = Operation.newBuilder().build()
        val listOfAllOperations = mutableListOf<Tezos.Operation>();

        if(isRevealed == false){
            val revealOperationData = RevealOperationData.newBuilder()
                .setPublicKey(ByteString.copyFrom(publicKey))

            revealOperation = Operation.newBuilder()
                .setSource(txData["source"] as String)
                .setFee(txData["reveal_fee"]!!.toLong())
                .setCounter(txData["reveal_counter"]!!.toLong())
                .setGasLimit(txData["reveal_gasLimit"]!!.toLong())
                .setStorageLimit(txData["reveal_storageLimit"]!!.toLong())
                .setKind(Operation.OperationKind.REVEAL)
                .setRevealOperationData(revealOperationData)
                .build()

            listOfAllOperations.add(revealOperation)
        }

        when (cmd) {
            "FA2" -> {
                val transferInfos = Txs.newBuilder()
                    .setAmount(txData["amount"] as String)
                    .setTokenId(txData["tokenId"] as String)
                    .setTo(txData["toAddress"] as String)
                    .build()

                val txObj = TxObject.newBuilder()
                    .setFrom(txData["senderAddress"] as String)
                    .addTxs(transferInfos)
                    .build()

                val fa2 = FA2Parameters.newBuilder()
                    .setEntrypoint("transfer")
                    .addTxsObject(txObj)
                    .build()

                val parameters = OperationParameters.newBuilder()
                    .setFa2Parameters(fa2)
                    .build()

                val transactionData = TransactionOperationData.newBuilder()
                    .setAmount(txData["transactionAmount"]!!.toLong())
                    .setDestination(txData["destination"] as String)
                    .setParameters(parameters)
                    .build()

                val transaction = Operation.newBuilder()
                    .setSource(txData["source"] as String)
                    .setFee(txData["fee"]!!.toLong())
                    .setCounter(txData["counter"]!!.toLong())
                    .setGasLimit(txData["gasLimit"]!!.toLong())
                    .setStorageLimit(txData["storageLimit"]!!.toLong())
                    .setKind(Operation.OperationKind.TRANSACTION)
                    .setTransactionOperationData(transactionData)
                    .build()

                listOfAllOperations.add(transaction)

                val operationList = OperationList.newBuilder()
                    .setBranch(txData["branch"] as String)
                    .addAllOperations(listOfAllOperations)
                    .build();

                    val signingInput = SigningInput.newBuilder()
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))
                    .setOperationList(operationList)
                    .build()

                val result = AnySigner.sign(signingInput, TEZOS, Tezos.SigningOutput.parser())
                txHash = Numeric.toHexString(result.encoded.toByteArray())
            }
            "FA12" -> {
                val fa12 = FA12Parameters.newBuilder()
                    .setEntrypoint("transfer")
                    .setFrom(txData["senderAddress"] as String)
                    .setTo(txData["toAddress"] as String)
                    .setValue(txData["value"] as String)
                    .build()

                val parameters = OperationParameters.newBuilder()
                    .setFa12Parameters(fa12)
                    .build()

                val transactionData = TransactionOperationData.newBuilder()
                    .setAmount(txData["transactionAmount"]!!.toLong())
                    .setDestination(txData["destination"] as String)
                    .setParameters(parameters)
                    .build()

                val transaction = Operation.newBuilder()
                    .setSource(txData["source"] as String)
                    .setFee(txData["fee"]!!.toLong())
                    .setCounter(txData["counter"]!!.toLong())
                    .setGasLimit(txData["gasLimit"]!!.toLong())
                    .setStorageLimit(txData["storageLimit"]!!.toLong())
                    .setKind(Operation.OperationKind.TRANSACTION)
                    .setTransactionOperationData(transactionData)
                    .build()

                listOfAllOperations.add(transaction)

                val operationList = OperationList.newBuilder()
                    .setBranch(txData["branch"] as String)
                    .addAllOperations(listOfAllOperations)
                    .build();

                val signingInput = SigningInput.newBuilder()
                    .setPrivateKey(ByteString.copyFrom(privateKey.data()))
                    .setOperationList(operationList)
                    .build()

                val result = AnySigner.sign(signingInput, TEZOS, Tezos.SigningOutput.parser())
                txHash = Numeric.toHexString(result.encoded.toByteArray())

            }
            else -> {
                val opJson = JSONObject(txData).toString()
                val result = AnySigner.signJSON(opJson, privateKey.data(), coinType!!.value())
                txHash = result
            }
        }
        return txHash
    }


}

File: ios/Classes/coins/XTZ.swift
/*
 XTZ
 */
import WalletCore

class XTZ: Coin  {
    init() {
        super.init(name: "XTZ", coinType: .tezos)
    }
    
    override func getPublicKey(path: String, mnemonic: String, passphrase: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeyEd25519().data.base64EncodedString()
    }
    
    override func getRawPublicKey(path: String, mnemonic: String, passphrase: String) -> FlutterStandardTypedData {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        return FlutterStandardTypedData(bytes: wallet!.getKey(coin: self.coinType, derivationPath: path).getPublicKeyEd25519().data)
    }
    
    override func signDataWithPrivateKey(path: String, mnemonic: String, passphrase: String, txData: String) -> String? {
        let wallet = HDWallet(mnemonic: mnemonic, passphrase: passphrase)
        let privateKey = wallet?.getKey(coin: self.coinType, derivationPath: path)
        let hash = Hash.blake2b(data: Data(hexString: txData)!, size: 32)
        if privateKey != nil {
            return privateKey?.sign(digest: hash, curve: self.coinType.curve)?.hexString
        }
        return nil
    }

    override func signTransaction(path: String, txData: [String : Any], mnemonic: String, passphrase: String) -> String? {
        var txHash: String? = nil
        let cmd = txData["cmd"] as! String?
        
        let privateKey = HDWallet(mnemonic: mnemonic, passphrase: passphrase)!.getKey(coin: self.coinType, derivationPath: path)
        let publicKey = privateKey.getPublicKeyEd25519().data
        
        switch(cmd){
            case "FA2":
                var operationList = TezosOperationList()
                operationList.branch = txData["branch"] as! String
                
                let transferInfos = TezosTxs.with{
                    $0.to = txData["toAddress"] as! String
                    $0.tokenID = txData["tokenId"] as! String
                    $0.amount = txData["amount"] as! String
                }
                
                let transactionOperationData = TezosTransactionOperationData.with {
                    $0.amount =  txData["transactionAmount"] as! Int64
                    $0.destination = txData["destination"] as! String
                    $0.parameters.fa2Parameters.entrypoint = "transfer";
                    $0.parameters.fa2Parameters.txsObject = [TezosTxObject.with{
                            $0.from = txData["senderAddress"] as! String
                            $0.txs = [transferInfos]
                        }]
                }

                var revealOperationData = TezosRevealOperationData()
                    revealOperationData.publicKey = publicKey

                var revealOperation = TezosOperation()
                    revealOperation.source = txData["source"] as! String;
                    revealOperation.fee = txData["fee"] as! Int64
                    revealOperation.counter = txData["counter"] as! Int64
                    revealOperation.gasLimit = txData["gasLimit"] as! Int64
                    revealOperation.storageLimit = txData["storageLimit"] as! Int64
                    revealOperation.kind = .reveal
                    revealOperation.revealOperationData = revealOperationData

                let transactionOperation = TezosOperation.with {
                    $0.source = txData["source"] as! String
                    $0.fee = txData["fee"] as! Int64
                    $0.counter = txData["counter"] as! Int64
                    $0.gasLimit = txData["gasLimit"] as! Int64
                    $0.storageLimit = txData["storageLimit"] as! Int64
                    $0.kind = .transaction
                    $0.transactionOperationData = transactionOperationData
                }
                
                operationList.operations = [ transactionOperation, revealOperation ]

                let input = TezosSigningInput.with {
                    $0.operationList = operationList
                    $0.privateKey = privateKey.data
                }

                let output: TezosSigningOutput = AnySigner.sign(input: input, coin: .tezos)

                txHash = output.encoded.hexString
            case "FA12": 
                let branch = txData["branch"] as! String
                var operationList = TezosOperationList()
                operationList.branch = branch
                
                let transactionOperationData = TezosTransactionOperationData.with {
                    $0.amount = txData["transactionAmount"] as! Int64
                    $0.destination = txData["destination"] as! String
                    $0.parameters.fa12Parameters.entrypoint = "transfer";
                    $0.parameters.fa12Parameters.from = txData["senderAddress"] as! String;
                    $0.parameters.fa12Parameters.to   = txData["toAddress"] as! String;
                    $0.parameters.fa12Parameters.value = txData["value"] as! String;
                }

                var revealOperationData = TezosRevealOperationData()
                    revealOperationData.publicKey = publicKey

                var revealOperation = TezosOperation()
                    revealOperation.source = txData["source"] as! String;
                    revealOperation.fee = txData["fee"] as! Int64
                    revealOperation.counter = txData["counter"] as! Int64
                    revealOperation.gasLimit = txData["gasLimit"] as! Int64
                    revealOperation.storageLimit = txData["storageLimit"] as! Int64
                    revealOperation.kind = .reveal
                    revealOperation.revealOperationData = revealOperationData
                
                let transactionOperation = TezosOperation.with {
                    $0.source = txData["source"] as! String;
                    $0.fee = txData["fee"] as! Int64
                    $0.counter = txData["counter"] as! Int64
                    $0.gasLimit = txData["gasLimit"] as! Int64
                    $0.storageLimit = txData["storageLimit"] as! Int64
                    $0.kind = .transaction
                    $0.transactionOperationData = transactionOperationData
                }
                
                operationList.operations = [ transactionOperation, revealOperation ]

                let input = TezosSigningInput.with {
                    $0.operationList = operationList
                    $0.privateKey = privateKey.data
                }

                let output: TezosSigningOutput = AnySigner.sign(input: input, coin: .tezos) 
                txHash = output.encoded.hexString

            default:
                let opJson =  Utils.objToJson(from: txData)
                let output =  AnySigner.signJSON(opJson!, key: privateKey.data, coin: self.coinType)
                txHash = output       
        }
        return txHash    
        
    }
    
}
